### PR TITLE
Prevents Indicator to Get Data With Same Timestamp

### PR DESCRIPTION
**Pull request by AlexCatarino**
#### Description
Prevents indicator to get data with the same timestamp. If the algorithm updates the indicator manually, it can use fill forward data, and that behavior is different from using helper methods (e.g. `StandardDeviation` vs `QCAlgorithm.STD`) since helper methods use consolidators and consolidators do not fill forward.

#### Motivation and Context
Standardize manual and automatic indicator updates. 

#### How Has This Been Tested?
```python
class StandardDeviation_test(QCAlgorithm):

    def Initialize(self):
        self.SetStartDate(2020, 1, 1) 
        self.spy = self.AddEquity("SPY", Resolution.Daily).Symbol
        self.SetWarmUp(101)

        self.std = StandardDeviation("SPY", 100) 
        self.std_short = self.STD("SPY", 100, Resolution.Daily)

    def OnData(self, data):
        self.std.Update(self.Time, self.Securities[self.spy].Close)
        if self.IsWarmingUp: return
        if self.std.Current.Value != self.std_short.Current.Value:
            raise Exception(f'Actual: {self.std.Current.Value} Expected: {self.std_short.Current.Value}')
```
#### Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Refactor (non-breaking change which improves implementation)
- [ ] Performance (non-breaking change which improves performance. Please add associated performance test and results)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Non-functional change (xml comments/documentation/etc)

#### Checklist:
- [x] My code follows the code style of this project.
- [x] I have read the **CONTRIBUTING** [document](https://github.com/QuantConnect/Lean/blob/master/CONTRIBUTING.md).
- [x] All new and existing tests passed.
- [x] My branch follows the naming convention `bug-<issue#>-<description>` or `feature-<issue#>-<description>`

### CODE REVIEW

File: Tests/Indicators/IndicatorExtensionsTests.cs [hunk 1]
```diff
@@ -35,9 +35,9 @@ public void PipesDataUsingOfFromFirstToSecond()
             second.Of(first);
 
             var data1 = new IndicatorDataPoint(DateTime.UtcNow, 1m);
-            var data2 = new IndicatorDataPoint(DateTime.UtcNow, 2m);
-            var data3 = new IndicatorDataPoint(DateTime.UtcNow, 3m);
-            var data4 = new IndicatorDataPoint(DateTime.UtcNow, 4m);
+            var data2 = new IndicatorDataPoint(DateTime.UtcNow.AddSeconds(1), 2m);
+            var data3 = new IndicatorDataPoint(DateTime.UtcNow.AddSeconds(2), 3m);
+            var data4 = new IndicatorDataPoint(DateTime.UtcNow.AddSeconds(3), 4m);
```
Comment (Martin-Molinero) [inline Tests/Indicators/IndicatorExtensionsTests.cs:40]: Looks like all these indicator tests should Not be using `DateTime.UtcNow` or `DateTime.Today` really but a fixed reference time

File: Tests/Indicators/RegressionChannelTest.cs [hunk 2]
```diff
@@ -40,10 +40,13 @@ public void ComputesCorrectly()
...
             for (var i = 0; i < prices.Length; i++)
             {
-                indicator.Update(time.AddMinutes(i), prices[i]);
-                stdDev.Update(time, prices[i]);
+                indicator.Update(time.AddMinutes(i + 1), prices[i]);
+                stdDev.Update(time.AddMinutes(i + 1), prices[i]);
                 actual[i] = Math.Round(indicator.Current.Value, 4);
             }
+
+            var s = string.Join(";", actual);
```
Comment (Martin-Molinero) [inline Tests/Indicators/RegressionChannelTest.cs:48]: nit - unused `s`?

File: Tests/Indicators/SharpeRatioTests.cs [hunk 3]
```diff
@@ -18,63 +18,66 @@
...
-    		// push the value 100000 into the indicator 20 times (sharpeRatioPeriod + movingAveragePeriod)
-    		for(int i = 0; i < 20; i++) {
-    			IndicatorDataPoint point = new IndicatorDataPoint(new DateTime(), 100000m);
-    			sr.Update(point);
-    		}
-    		
-    		Assert.AreEqual(sr.Current.Value, 0m);
+            // With the value not changing, the indicator should return default value 0m.
+            var sharpeRatio = new SharpeRatio("SR", 10);
+            var time = DateTime.Now;
```
Comment (Martin-Molinero) [inline Tests/Indicators/SharpeRatioTests.cs:40]: Same comment about `DateTime.Now`, think it's clearer and probably faster if just using a reference time

File: Tests/TestData/spy_rvi.txt [hunk 4]
```diff
@@ -1,380 +1,380 @@
...
+7/23/2019 01:30 PM,299.14,300.03,298.22,300.03,44657973,51649971.8,-0.07118005652694373,0.04779468963618085
+7/24/2019 01:30 PM,299.19,301.44,299.09,301.44,47213232,49909200.6,-0.09387153010594568,-0.02508510254440847
+7/25/2019 01:30 PM,300.94,301,299.11,300,55394092,50099660.1,-0.07488866672726177,-0.06793819577467955
+7/26/2019 01:30 PM,300.76,302.23,300.62,302.01,45084077,50336102,-0.05697726259474461,-0.07761295213135054
+7/29/2019 01:30 PM,301.88,301.93,300.85,301.46,38126462,49274878.65,-0.046962290502798286,-0.06742761320879279
+7/30/2019 01:30 PM,299.91,301.17,299.49,300.72,45849045,47611957.2,-0.032138867405693376,-0.05248444005467349
+7/31/2019 01:30 PM,300.99,301.2,295.2,297.43,104245162,49748988.1,-0.01170361823127466,-0.0378138661071671
+8/1/2019 01:30 PM,297.6,300.87,293.96,294.84,142646576,54836372.1,-0.04076756514597638,-0.029235804487118457
+8/2/2019 01:30 PM,293.85,294.12,290.9,292.62,116749695,58089990.55,-0.0927499668038819,-0.038305200160679564
+8/5/2019 01:30 PM,288.09,288.21,281.72,283.82,178745449,64735174.4,-0.15201900237530114,-0.07179294741771539
```
Comment (Martin-Molinero) [inline Tests/TestData/spy_rvi.txt:35]: Think these changes look culture-specific? Probably some bug with writing them to file, should be culture invariant?

File: Indicators/IndicatorBase.cs [hunk 5]
```diff
@@ -77,7 +77,7 @@ protected IndicatorBase(string name)
         public bool Update(IBaseData input)
         {
             T _previousSymbolInput = default(T);
-            if (_previousInput.TryGetValue(input.Symbol.ID, out _previousSymbolInput) && input.EndTime < _previousSymbolInput.EndTime)
+            if (_previousInput.TryGetValue(input.Symbol.ID, out _previousSymbolInput) && input.EndTime <= _previousSymbolInput.EndTime)
```
Comment (Martin-Molinero) [inline Indicators/IndicatorBase.cs:80]: Interesting change! We should probably also add some new test covering this new behavior 👍

**Review by Martin-Molinero — COMMENTED** [review]
Cool!
Wonder what issue triggered this adjustment, multi-resolution algorithms? because fill forwarded data end time should still be advanced?

File: Tests/TestData/spy_rvi.txt (hunk 4 above)
Comment (AlexCatarino) [reply Tests/TestData/spy_rvi.txt:35]: Yes, these are culture-specific, but it respects the date/time format of other files used for indicator validation. I think it's simpler to change the file for the scope of this fix.

File: Indicators/IndicatorBase.cs (hunk 5 above)
Comment (AlexCatarino) [reply Indicators/IndicatorBase.cs:80]: I added this test. Fails on master, passes in this PR:
https://github.com/QuantConnect/Lean/pull/5442/files#diff-027dc7219a991f6c6abddceb033c0e0a2e841c166553845e0766b0d04403c18bR96

**Comment by jaredbroad** [comment]
Suggest we merge once the root bug of the duplicate bar from the end of warm-up is addressed to not hide the warm-up issue.

**Review by jaredbroad — CHANGES_REQUESTED** [review]
Marking request changes so we can get to bottom of why there's duplicated data.

**Comment by Martin-Molinero** [comment]
Closing for now, related underlying issue https://github.com/QuantConnect/Lean/issues/4048

**Pull request closed without merging**